java/vector/src/main/java/org/apache/arrow/vector/stream/MessageSerializer.java

```diff
@@ -98,26 +100,41 @@ public static Schema deserializeSchema(ReadChannel in) throws IOException {
   }
 
   /**
-   * Serializes an ArrowRecordBatch.
+   * Serializes an ArrowRecordBatch. Returns the offset and length of the written batch.
    */
-  public static long serialize(WriteChannel out, ArrowRecordBatch batch)
+  public static ArrowBlock serialize(WriteChannel out, ArrowRecordBatch batch)
       throws IOException {
     long start = out.getCurrentPosition();
     int bodyLength = batch.computeBodyLength();
-
     ByteBuffer metadata = WriteChannel.serialize(batch);
+
+    int messageLength = 4 + metadata.remaining() + bodyLength;
     ByteBuffer serializedHeader =
-        serializeHeader(MessageHeader.RecordBatch, bodyLength + metadata.remaining() + 4);
+        serializeHeader(MessageHeader.RecordBatch, messageLength);
```

Comment:
The serialization would be quite a bit simpler if we enforced that Message was also fixed length. It would be similarly easy if we enforced that the current fields in Message always came first.
i.e. read(sizeof(Message)) was always going to be valid and if the header grew, you could tell by looking header.version.